app/models/manufacturer.rb

```diff
@@ -21,15 +21,18 @@ class Manufacturer < ApplicationRecord
 
   scope :alphabetized, -> { order(:name) }
 
-  def volume
-    # returns 0 instead of nil when Manufacturer exists without any donations
-    donations.joins(:line_items).sum(:quantity)
-  end
-
-  def self.by_donation_count(count = 10)
-    # selects manufacturers that have donation qty > 0
+  def self.by_donation_count(count = 10, date_range = nil)
+    # selects manufacturers that have donation qty > 0 in the provided date range
     # and sorts them by highest volume of donation
-    select { |m| m.volume.positive? }.sort.reverse.first(count)
+    manufacturers = select do |m|
+      donation_volume = m.donations.joins(:line_items).where(issued_at: date_range).sum(:quantity)
+      if donation_volume.positive?
+        m.instance_variable_set(:@num_of_donations, donation_volume)
```

Comment:
We really don't want to set instance variables from outside the class. The way I'd do this is to create a method on `Manufacturer` called `donation_count` which you can memoize if necessary by setting the result to an instance variable if it doesn't exist, e.g. `@donation_count ||= ...` .

Having said that, this is something you can solve in straight SQL (or ActiveRecord). It should be something along the lines of:

```ruby
self.joins(donations: :line_items).where(donations: {issued_at: date_range}).
   select('sum(quantity) as quantity_count').
   group('manufacturers.id').
   having('sum(quantity) > 0').
   order('quantity_count desc').to_a
```

https://chatgpt.com/g/g-m5lMeGifF-sql-expert is a good resource to talk through this if necessary.

Reply:
That looks like a significant enough adjustment that I'm going to hold off on the final functional check.